r/src/arrow_types.h

```diff
@@ -37,25 +37,74 @@ struct NoDelete {
   inline void operator()(T* ptr){};
 };
 
+namespace arrow {
+namespace r {
+struct symbols {
+  static SEXP units;
+  static SEXP xp;
+};
+}  // namespace r
+}  // namespace arrow
+
 namespace Rcpp {
+namespace internal {
+
+template <typename Pointer>
+Pointer r6_to_smart_pointer(SEXP self) {
+  return reinterpret_cast<Pointer>(
+      EXTPTR_PTR(Rf_findVarInFrame(self, arrow::r::symbols::xp)));
+}
+
+}  // namespace internal
+
+template <typename T>
+class ConstReferenceSmartPtrInputParameter {
```

Comment:
Should this be called "UnwrapExternalPointer" or something?

Reply:
This is following the existing pattern from Rcpp, that has `ConstReferenceInputParameter<T>`